App binds to 127.0.0.1:3000 by default

adamlamar opened the pull request:
In `options.go`, the `Options.Host` field states the default bind location is `127.0.0.1:3000`. However, this value doesn't seem to be used anywhere in buffalo.

I noticed that I was presented with Mac's firewall prompt every time my buffalo app started. This prompt isn't presented when a program binds to `127.0.0.1` only, so the app must be binding to `0.0.0.0`.

<img width="418" alt="firewall" src="https://user-images.githubusercontent.com/13948507/31857992-d2ae2cea-b6a9-11e7-988d-d5924b759c9c.png">

Rather than use the bind address/port provided to `app.Start()`, this PR adjusts `app.Start()` to use the `Options.Host` value. This should provide the intended behavior from the `Options.Host` description.

It also removes the argument to `app.Start()`, which would be an issue for app upgrades where `main.go` was generated to provide the `addr` argument. 

After building a test project with this change, the app binds to `127.0.0.1:3000` by default. `PORT` can still be overridden as before, and the Mac's firewall prompt no longer pops on app start.

markbates commented:
Hi, thanks for the PR. A couple of things, first, all PR's need to be against the `development` branch, as noted in the CONTRIBUTING.md file. I've changed the branch, but now there are conflicts that need resolution. 

Second the tests on Travis are failing, so those need to be addressed. You can run the full set of integration tests locally, with `docker build .`.

Finally, the `Host` option is meant to be the "public" host, so things like `https://foo.bar`, not the host to bind to. By default, Go, does bind to `0.0.0.0` if there is not host present. I'm up for setting the ability to change that, but I think we need a different option for that, perhaps something like `ServerHost` or `BindHost`, something along those lines. Thoughts?

markbates commented:
I like Addr. That’s good. Let’s use that.

By public url I meant the actual address of their site.

-----------
Mark Bates

On Oct 23, 2017, 7:34 PM -0400, Adam Lamar <notifications@github.com>, wrote:
> Whoops, I completely missed the CONTRIBUTING.md file - will go over that and the failing tests.
> It seems like I misunderstood the purpose of the Host as well. Could you clarify what you mean by "public" host? Is it meant for virtual hosting, or building absolute links in HTML templates, or something else?
> As for a separate option, that sounds good to me. I'd prefer something like BindAddr or maybe just Addr since its passing the value into http.Server.Addr and that would potentially be familiar to other Go devs.
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

adamlamar commented:
@markbates I rebased this branch against `development`, force-pushed to my fork's branch and addressed the merge conflicts. Also used `Addr` instead of `Host`. Mind taking another look?

markbates commented:
I created a new PR for this #719 because I wanted to create some deprecations around the `Start` function in particular, since it was a breaking change. Hope you don't mind.
